### ISSUE TITLE
Support "SVDiff: Compact Parameter Space for Diffusion Fine-Tuning"

### ISSUE THREAD

**Issue by mkshing** [huggingface/diffusers#3071]
### Summary
[SVDiff: Compact Parameter Space for Diffusion Fine-Tuning](https://arxiv.org/abs/2303.11305) is an effective domain-tuning method for text-to-image models like LoRA but the number of trainable extra parameters is less than LoRA and the result is better or compatible. 

If you don't have time to read the paper, please refer to [my summary tweet](https://github.com/mkshing/svdiff-pytorch) for more details.

I have already implemented it as an unofficial one and the result is very promising. And, @sayakpaul and @AK391 suggested me to integrate into diffusers! (Apparently, the original code will be released about in 2 weeks so I am totally fine if you want to wait for the original one or not :))

- github: https://github.com/mkshing/svdiff-pytorch
- gradio: https://huggingface.co/spaces/svdiff-library/SVDiff-Training-UI
- svdiff org in Hub: https://huggingface.co/svdiff-library

Before send a PR, I wanted to discuss about a couple of things. 

![image](https://user-images.githubusercontent.com/33302880/231468112-49c409c6-0c68-4115-be81-26a4779db675.png)

### Discussion
To raise my discussion topic, please let me explain the background briefly. 

In SVDiff, you need to get the updated weight by a certain function g at every step like below.
```
W' = g(W, \delta)
```
where,` \delta` is an extra parameters SVDiff wants to optimize and `W`  is frozen.
So, if you have a linear layer w/o biases, `Y = W * X` would be `Y = W' * X` instead.

In my current implementation, I manually replaced `nn.Linear` to the custom `SVDLinear` which is inherited from `nn.Linear`.

```
class SVDLinear(nn.Linear):
  def __init__(self, in_features, out_features, **kwargs):
    nn.Linear.__init__(self, in_features, out_features, **kwargs)
    self.delta = nn.Parameter(torch.zeros(in_features))
    
  def function_g(self, w, delta):
      ### return updated weights ####

  def forward(self, x):
     weight_updated = self.function_g(self.weight, self.delta)
     return F.linear(x, weight_updated, bias=self.bias)
```
In this case, **you need to replace all nn.Linear to SVDLinear manually in all files related to the t2i model.**
Additionally, the actual case is you need to replace `nn.Linear`, `nn.Conv1D`, `nn.Conv2D`, etc to custom classes defined [here](https://github.com/mkshing/svdiff-pytorch/blob/main/svdiff_pytorch/layers.py).

So, I believe that we can't use AttnProcesso like `LoRAAttnProcessor` in this case and I need you to discuss what is the clever way to integrate it. 

FYI, actually, I got feedback from the author directly and they recommend to use PyTorch's [parametrization](https://pytorch.org/tutorials/intermediate/parametrizations.html) functionality to register hooks on loaded models. 

### Citation 
I would like to thank @phymhan for the amazing paper. 
```
@misc{https://doi.org/10.48550/arXiv.2303.11305,
      title         = {SVDiff: Compact Parameter Space for Diffusion Fine-Tuning}, 
      author        = {Ligong Han and Yinxiao Li and Han Zhang and Peyman Milanfar and Dimitris Metaxas and Feng Yang},
      year          = {2023},
      eprint        = {2303.11305},
      archivePrefix = {arXiv},
      primaryClass  = {cs.CV},
      url           = {https://arxiv.org/abs/2303.11305}
}
```

## Open source status

- [X] The model implementation is available
- [X] The model weights are available (Only relevant if addition is not a scheduler).

**Comment by sayakpaul** [comment]
Thanks so much for putting this together!

> FYI, actually, I got feedback from the author directly and they recommend to use PyTorch's [parametrization](https://pytorch.org/tutorials/intermediate/parametrizations.html) functionality to register hooks on loaded models.

IIUC if we go via the parameterization route, we won't have to do `SVD-*` classes, no? We could, for example, register them as hooks on loaded models as you mentioned. Should we explore that route instead?

**Comment by mkshing** [comment]
Yes, you're correct! Actually, I did try this way quickly but I got stuck on some problems around `requires_grad`. So, I would like someone to help or give me some examples. 
@phymhan do you think it's possible to provide an example script??

**Comment by sayakpaul** [comment]
> Actually, I did try this way quickly but I got stuck on some problems around requires_grad.

Could you maybe provide some snippets so that we can take a look?

**Comment by phymhan** [comment]
Hi @mkshing and @sayakpaul , sure here is a simple snippet
https://github.com/phymhan/SVDiff/blob/4ca843cde194fb3739659356435582b16392402e/utils_svd.py#L14
for every module we wish to optimize, we register a new `_SpectralShift` module to it:
https://github.com/phymhan/SVDiff/blob/4ca843cde194fb3739659356435582b16392402e/utils_svd.py#L97
and only put parameters in `new_module` in the optimizer. Thanks!

**Comment by mkshing** [comment]
@phymhan @sayakpaul thank you for sharing the example code! I have made a PR finally! 
https://github.com/huggingface/diffusers/pull/3107
